packages/web/src/common/store/cache/tracks/utils/fetchTrackStreamUrls.ts

```diff
@@ -0,0 +1,72 @@
+import { ID } from '@audius/common/models'
+import {
+  cacheTracksActions,
+  getContext,
+  accountSelectors,
+  gatedContentSelectors,
+  cacheTracksSelectors
+} from '@audius/common/store'
+import { getQueryParams } from '@audius/common/utils'
+import { all, call, select, put } from 'typed-redux-saga'
+const { getUserId } = accountSelectors
+const { getTrackStreamUrl } = cacheTracksSelectors
+const { setStreamUrl } = cacheTracksActions
+const { getNftAccessSignatureMap } = gatedContentSelectors
+
+export function* fetchTrackStreamUrls({ trackIds }: { trackIds: ID[] }) {
+  const apiClient = yield* getContext('apiClient')
+  const audiusBackendInstance = yield* getContext('audiusBackendInstance')
+  const reportToSentry = yield* getContext('reportToSentry')
+
+  const currentUserId = yield* select(getUserId)
+
+  try {
+    // TODO: Ideally we should batch this method (needs a backend change to support)
+    const callEffects = trackIds.map(function* (id) {
+      try {
+        const existingUrl = yield* select(getTrackStreamUrl, id)
+
+        if (existingUrl) {
+          return existingUrl
+        }
+
+        const nftAccessSignatureMap = yield* select(getNftAccessSignatureMap)
+        const nftAccessSignature = nftAccessSignatureMap[id]?.mp3 ?? null
+        const queryParams = yield* call(getQueryParams, {
+          audiusBackendInstance,
+          nftAccessSignature
+        })
+        const streamUrl = yield* call([apiClient, 'getTrackStreamUrl'], {
+          id,
+          currentUserId,
+          queryParams,
+          abortOnUnreachable: true
+        })
+        if (streamUrl !== undefined) {
+          // Set the stream url in the cache
+          yield* put(setStreamUrl(id, streamUrl))
```

Comment:
would we want a batch set in state so we dont have n number of state updates?

Reply:
imo the effects should call the api and return the urls, then after the `all` call we call `put` with the array of urls? the array i guess needs {id, streamUrl} in that case. anyway an idea

Reply:
Good point, in my head I was thinking since we're fetching independently, why not put them independently so they're ready to stream one by one. 
But I think that's probably not necessary for one, and probably counter-intuitive performance-wise retriggering redux logic so often

Reply:
yeah since this add-to-cache call is a batch like operation, lets keep the stream url update as a batch thing as well

Reply:
Right, and ideally the backend would also support a batch call so that everything is batched